File: TarkovDevAPI/TarkovDevAPI.py
import sqlite3
import requests
import os

SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))
DB_PATH = os.path.join(SCRIPT_DIR, 'EFT_Data.db')

def run_query(query):
    headers = {"Content-Type": "application/json"}
    response = requests.post('https://api.tarkov.dev/graphql', headers=headers, json={'query': query})
    if response.status_code == 200:
        return response.json()
    else:
        raise Exception("Query failed to run by returning code of {}. {}".format(response.status_code, query))

ItemQuery = """
{
  items(lang: en) {
    id
    sellFor {
      priceRUB
      vendor {
        name
      }
    }
    shortName
  }
}
"""
def UpdateItemTable():
    con = sqlite3.connect(DB_PATH)
    cur = con.cursor()
    result = run_query(ItemQuery)
    for Item in result['data']['items']:

        HighestTraderPrice = -1
            
        for SellOption in Item['sellFor']:
            price = SellOption['priceRUB']
            vendor = SellOption["vendor"]["name"]
            if(vendor == "Flea Market"):
                continue

            if(price > HighestTraderPrice):
                HighestTraderPrice = price

        cur.execute("INSERT OR IGNORE INTO item_data (bsg_id, short_name, trader_price) VALUES (?, ?, ?)", (Item['id'], Item['shortName'], HighestTraderPrice))

    con.commit()
    con.close()
    return

ContainerQuery = """
{
  lootContainers {
    id
    name
  }
}
"""
def UpdateContainerTable():
    con = sqlite3.connect(DB_PATH)
    cur = con.cursor()

    result = run_query(ContainerQuery)

    for Item in result['data']['lootContainers']:
       cur.execute("INSERT OR IGNORE INTO container_data (bsg_id, short_name) VALUES (?, ?)", (Item['id'], Item['name']))

    con.commit()
    con.close()
    return

AmmoQuery ="""
{
  ammo {
    item {
      id
      shortName
    }
  }
}
"""
def UpdateAmmoTable():
    result = run_query(AmmoQuery)

    con = sqlite3.connect(DB_PATH)
    cur = con.cursor()

    for Item in result['data']['ammo']:
         cur.execute("INSERT OR IGNORE INTO ammo_data (bsg_id, short_name) VALUES (?, ?)", (Item['item']['id'], Item['item']['shortName']))

    con.commit()
    con.close()
    return

ExfilQuery = """
{
  maps {
    nameId
    extracts {
      name
      position {
        x
        y
        z
      }
    }
  }
}
"""
def UpdateExfilTable():
    """Fetch exfil data from tarkov.dev and update the database with positions"""
    
    MAP_NAME_MAPPING = {
        'Woods': ['woods', 'Woods'],
        'Shoreline': ['shoreline', 'Shoreline'],
        'RezervBase': ['rezervbase', 'RezervBase'],
        'Labyrinth': ['Labyrinth'],
        'laboratory': ['laboratory'],
        'Interchange': ['interchange', 'Interchange'],
        'factory4_day': ['factory4_day'],
        'factory4_night': ['factory4_night'],
        'bigmap': ['bigmap'],
        'Lighthouse': ['lighthouse', 'Lighthouse'],
        'TarkovStreets': ['tarkovstreets', 'TarkovStreets'],
        'Sandbox': ['Sandbox'],
        'Sandbox_high': ['Sandbox_high'],
        'Terminal': ['Terminal'],
        'Sandbox_start': ['Sandbox_start'],
    }
    
    con = sqlite3.connect(DB_PATH)
    cur = con.cursor()
    
    result = run_query(ExfilQuery)
    
    total_inserted = 0
    for map_data in result['data']['maps']:
        map_api_id = map_data['nameId']
        internal_names = MAP_NAME_MAPPING.get(map_api_id, [])
        
        if not internal_names:
            continue
        
        for extract in map_data['extracts']:
            pos = extract.get('position', {})
            pos_x = pos.get('x', 0.0)
            pos_y = pos.get('y', 0.0)
            pos_z = pos.get('z', 0.0)
            
            for internal_name in internal_names:
                cur.execute("""
                    INSERT OR REPLACE INTO exfil_data 
                    (map_internal_name, exfil_display_name, pos_x, pos_y, pos_z) 
                    VALUES (?, ?, ?, ?, ?)
                """, (internal_name, extract['name'], pos_x, pos_y, pos_z))
                total_inserted += 1
    
    con.commit()
    con.close()

if __name__ == "__main__":
    print(f"Using database: {DB_PATH}")
    UpdateAmmoTable()
    UpdateItemTable()
    UpdateContainerTable()
    UpdateExfilTable()

File: CyNickal Software EFT/Database/TarkovDevAPI.cpp
#include "pch.h"
#include "TarkovDevAPI.hpp"
#include "Network/Classes/CGraphQLRequest/CGraphQLRequest.hpp"

void TarkovDevAPI::FetchAll(sqlite3* db)
{
	std::println("[TarkovDevAPI] Fetching latest data from tarkov.dev...");
	FetchItems(db);
	FetchContainers(db);
	FetchAmmo(db);
	FetchExfils(db);
	std::println("[TarkovDevAPI] Fetch complete.");
}

void TarkovDevAPI::FetchItems(sqlite3* db)
{
	constexpr const char* query = R"(
	{
	  items(lang: en) {
	    id
	    sellFor {
	      priceRUB
	      vendor {
	        name
	      }
	    }
	    shortName
	  }
	}
	)";

	CGraphQLRequest req(API_URL, query);
	if (req.GetResponse().empty())
	{
		std::println("[TarkovDevAPI] FetchItems: empty response, skipping.");
		return;
	}

	auto json = nlohmann::json::parse(req.GetResponse(), nullptr, false);
	if (json.is_discarded() || !json.contains("data") || !json["data"].contains("items"))
	{
		std::println("[TarkovDevAPI] FetchItems: failed to parse response.");
		return;
	}

	sqlite3_exec(db, "BEGIN TRANSACTION;", nullptr, nullptr, nullptr);

	sqlite3_stmt* stmt = nullptr;
	sqlite3_prepare_v2(db,
		"INSERT OR IGNORE INTO item_data (bsg_id, short_name, trader_price) VALUES (?, ?, ?);",
		-1, &stmt, nullptr);

	for (const auto& item : json["data"]["items"])
	{
		int highestTraderPrice = -1;
		for (const auto& sell : item["sellFor"])
		{
			std::string vendor = sell["vendor"]["name"].get<std::string>();
			if (vendor == "Flea Market")
				continue;
			int price = sell["priceRUB"].get<int>();
			if (price > highestTraderPrice)
				highestTraderPrice = price;
		}

		std::string id = item["id"].get<std::string>();
		std::string shortName = item["shortName"].get<std::string>();

		sqlite3_bind_text(stmt, 1, id.c_str(), -1, SQLITE_TRANSIENT);
		sqlite3_bind_text(stmt, 2, shortName.c_str(), -1, SQLITE_TRANSIENT);
		sqlite3_bind_int(stmt, 3, highestTraderPrice);
		sqlite3_step(stmt);
		sqlite3_reset(stmt);
	}

	sqlite3_finalize(stmt);
	sqlite3_exec(db, "COMMIT;", nullptr, nullptr, nullptr);
	std::println("[TarkovDevAPI] FetchItems: done.");
}

void TarkovDevAPI::FetchContainers(sqlite3* db)
{
	constexpr const char* query = R"(
	{
	  lootContainers {
	    id
	    name
	  }
	}
	)";

	CGraphQLRequest req(API_URL, query);
	if (req.GetResponse().empty())
	{
		std::println("[TarkovDevAPI] FetchContainers: empty response, skipping.");
		return;
	}

	auto json = nlohmann::json::parse(req.GetResponse(), nullptr, false);
	if (json.is_discarded() || !json.contains("data") || !json["data"].contains("lootContainers"))
	{
		std::println("[TarkovDevAPI] FetchContainers: failed to parse response.");
		return;
	}

	sqlite3_exec(db, "BEGIN TRANSACTION;", nullptr, nullptr, nullptr);

	sqlite3_stmt* stmt = nullptr;
	sqlite3_prepare_v2(db,
		"INSERT OR IGNORE INTO container_data (bsg_id, short_name) VALUES (?, ?);",
		-1, &stmt, nullptr);

	for (const auto& container : json["data"]["lootContainers"])
	{
		std::string id = container["id"].get<std::string>();
		std::string name = container["name"].get<std::string>();

		sqlite3_bind_text(stmt, 1, id.c_str(), -1, SQLITE_TRANSIENT);
		sqlite3_bind_text(stmt, 2, name.c_str(), -1, SQLITE_TRANSIENT);
		sqlite3_step(stmt);
		sqlite3_reset(stmt);
	}

	sqlite3_finalize(stmt);
	sqlite3_exec(db, "COMMIT;", nullptr, nullptr, nullptr);
	std::println("[TarkovDevAPI] FetchContainers: done.");
}

void TarkovDevAPI::FetchAmmo(sqlite3* db)
{
	constexpr const char* query = R"(
	{
	  ammo {
	    item {
	      id
	      shortName
	    }
	  }
	}
	)";

	CGraphQLRequest req(API_URL, query);
	if (req.GetResponse().empty())
	{
		std::println("[TarkovDevAPI] FetchAmmo: empty response, skipping.");
		return;
	}

	auto json = nlohmann::json::parse(req.GetResponse(), nullptr, false);
	if (json.is_discarded() || !json.contains("data") || !json["data"].contains("ammo"))
	{
		std::println("[TarkovDevAPI] FetchAmmo: failed to parse response.");
		return;
	}

	sqlite3_exec(db, "BEGIN TRANSACTION;", nullptr, nullptr, nullptr);

	sqlite3_stmt* stmt = nullptr;
	sqlite3_prepare_v2(db,
		"INSERT OR IGNORE INTO ammo_data (bsg_id, short_name) VALUES (?, ?);",
		-1, &stmt, nullptr);

	for (const auto& ammo : json["data"]["ammo"])
	{
		std::string id = ammo["item"]["id"].get<std::string>();
		std::string shortName = ammo["item"]["shortName"].get<std::string>();

		sqlite3_bind_text(stmt, 1, id.c_str(), -1, SQLITE_TRANSIENT);
		sqlite3_bind_text(stmt, 2, shortName.c_str(), -1, SQLITE_TRANSIENT);
		sqlite3_step(stmt);
		sqlite3_reset(stmt);
	}

	sqlite3_finalize(stmt);
	sqlite3_exec(db, "COMMIT;", nullptr, nullptr, nullptr);
	std::println("[TarkovDevAPI] FetchAmmo: done.");
}

void TarkovDevAPI::FetchExfils(sqlite3* db)
{
	constexpr const char* query = R"(
	{
	  maps {
	    nameId
	    extracts {
	      name
	      position {
	        x
	        y
	        z
	      }
	    }
	  }
	}
	)";

	// Maps tarkov.dev nameId -> list of internal game map names
	static const std::unordered_map<std::string, std::vector<std::string>> MAP_NAME_MAPPING = {
		{ "Woods",       { "woods", "Woods" } },
		{ "Shoreline",   { "shoreline", "Shoreline" } },
		{ "RezervBase",  { "rezervbase", "RezervBase" } },
		{ "Labyrinth",   { "Labyrinth" } },
		{ "laboratory",  { "laboratory" } },
		{ "Interchange", { "interchange", "Interchange" } },
		{ "factory4_day",   { "factory4_day" } },
		{ "factory4_night", { "factory4_night" } },
		{ "bigmap",      { "bigmap" } },
		{ "Lighthouse",  { "lighthouse", "Lighthouse" } },
		{ "TarkovStreets", { "tarkovstreets", "TarkovStreets" } },
		{ "Sandbox",       { "Sandbox" } },
		{ "Sandbox_high",  { "Sandbox_high" } },
		{ "Terminal",      { "Terminal" } },
		{ "Sandbox_start", { "Sandbox_start" } },
	};

	CGraphQLRequest req(API_URL, query);
	if (req.GetResponse().empty())
	{
		std::println("[TarkovDevAPI] FetchExfils: empty response, skipping.");
		return;
	}

	auto json = nlohmann::json::parse(req.GetResponse(), nullptr, false);
	if (json.is_discarded() || !json.contains("data") || !json["data"].contains("maps"))
	{
		std::println("[TarkovDevAPI] FetchExfils: failed to parse response.");
		return;
	}

	sqlite3_exec(db, "BEGIN TRANSACTION;", nullptr, nullptr, nullptr);

	sqlite3_stmt* stmt = nullptr;
	sqlite3_prepare_v2(db,
		"INSERT OR REPLACE INTO exfil_data (map_internal_name, exfil_display_name, pos_x, pos_y, pos_z) VALUES (?, ?, ?, ?, ?);",
		-1, &stmt, nullptr);

	for (const auto& map : json["data"]["maps"])
	{
		std::string nameId = map["nameId"].get<std::string>();

		auto it = MAP_NAME_MAPPING.find(nameId);
		if (it == MAP_NAME_MAPPING.end())
			continue;

		for (const auto& extract : map["extracts"])
		{
			std::string extractName = extract["name"].get<std::string>();
			double posX = extract["position"]["x"].get<double>();
			double posY = extract["position"]["y"].get<double>();
			double posZ = extract["position"]["z"].get<double>();

			for (const auto& internalName : it->second)
			{
				sqlite3_bind_text(stmt, 1, internalName.c_str(), -1, SQLITE_TRANSIENT);
				sqlite3_bind_text(stmt, 2, extractName.c_str(), -1, SQLITE_TRANSIENT);
				sqlite3_bind_double(stmt, 3, posX);
				sqlite3_bind_double(stmt, 4, posY);
				sqlite3_bind_double(stmt, 5, posZ);
				sqlite3_step(stmt);
				sqlite3_reset(stmt);
			}
		}
	}

	sqlite3_finalize(stmt);
	sqlite3_exec(db, "COMMIT;", nullptr, nullptr, nullptr);
	std::println("[TarkovDevAPI] FetchExfils: done.");
}
